backend/app/adapter/db/userrole.go

```diff
@@ -0,0 +1,213 @@
+	}
+
+	return decodeRoles(encoded), nil
+}
+
+func (u UserRoleSQL) AddRole(user entity.User, r role.Role) error {
+	return u.saveRoles(user, []role.Role{r}, nil)
+}
+
+func (u UserRoleSQL) DeleteRole(user entity.User, r role.Role) error {
+	return u.saveRoles(user, nil, []role.Role{r})
+}
+
+func NewUserRoleSQL(db *sql.DB) UserRoleSQL {
+	return UserRoleSQL{db: db}
+}
+
+func (u UserRoleSQL) saveRoles(user entity.User, toAdd, toDelete []role.Role) error {
+	tx, err := u.db.Begin()
+
+	if err != nil {
+		return nil
+	}
+
+	defer tx.Rollback()
+
+	query := fmt.Sprintf(`
+SELECT "%s" 
+FROM "%s" 
+WHERE "%s"=$1 FOR UPDATE;
+`,
+		table.UserRole.ColumnRoles,
+		table.UserRole.TableName,
+		table.UserRole.ColumnUserID,
+	)
+
+	encoded := ""
+	err = tx.QueryRow(query, user.ID).Scan(&encoded)
+	isNew := err == sql.ErrNoRows
+
+	if err != nil && !isNew {
+		return err
+	}
+
+	roles := modifyRoles(decodeRoles(encoded), toAdd, toDelete)
+
+	if len(roles) > 0 {
+		if isNew {
+			if err := insertExec(tx, user.ID, roles); err != nil {
+				return err
+			}
+		} else {
+			if err := updateExec(tx, user.ID, roles); err != nil {
+				return err
+			}
+		}
+	} else {
+		if err := deleteExec(tx, user.ID); err != nil {
+			return err
+		}
+	}
+
+	if err := tx.Commit(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func insertExec(tx *sql.Tx, userID string, roles []role.Role) error {
+	statement := fmt.Sprintf(`
+INSERT INTO "%s" ("%s", "%s")
+VALUES ($1, $2);
+`,
+		table.UserRole.TableName,
+		table.UserRole.ColumnUserID,
+		table.UserRole.ColumnRoles,
+	)
+
+	if _, err := tx.Exec(statement, userID, encodeRoles(roles)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func updateExec(tx *sql.Tx, userID string, roles []role.Role) error {
+	statement := fmt.Sprintf(`
+UPDATE "%s"
+SET "%s"=$1
+WHERE "%s"=$2;
+`,
+		table.UserRole.TableName,
+		table.UserRole.ColumnRoles,
+		table.UserRole.ColumnUserID,
+	)
+
+	if _, err := tx.Exec(statement, encodeRoles(roles), userID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func deleteExec(tx *sql.Tx, userID string) error {
+	statement := fmt.Sprintf(`
+DELETE FROM "%s"
+WHERE "%s"=$1;
+`,
+		table.UserRole.TableName,
+		table.UserRole.ColumnUserID,
+	)
+
+	if _, err := tx.Exec(statement, userID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func modifyRoles(roles []role.Role, toAdd, toDelete []role.Role) []role.Role {
+	for _, add := range toAdd {
+		roles = append(roles, add)
+	}
+
+	for _, delete := range toDelete {
+		for i, r := range roles {
+			if r == delete {
+				roles[i], roles[len(roles)-1] = roles[len(roles)-1], roles[i]
+				roles = roles[:len(roles)-1]
+				break
+			}
+		}
+	}
+
+	return roles
+}
+
+func encodeRoles(roles []role.Role) string {
+	if len(roles) == 0 {
+		return ""
+	}
+
+	sort.Slice(roles, func(i, j int) bool {
+		return roles[i] > roles[j]
+	})
+
+	var encoded string
+
+	for r := roles[0]; r >= 0; r-- {
+		if len(roles) > 0 && r == roles[0] {
+			encoded += "1"
```

Comment:
Since the roles are stored in the database as array of characters, can we just append the character of the role for example if role = 2 we'd append string(2) and the decoding is done in a similar way?

Reply:
It is not going to work. How to handle the next situation:
```
RoleA = 1
RoleB = 2
...
RoleC = 21
RoleD = 22
```

Simple append doesn't work here:)

Reply:
@arberiii We have to do shift and masking. Each bit represents whether an role is assigned an user. 1 means yes, 0 means no.

@alldroll Please correct me if I am wrong. I saw in Kernel, lots of configuration management has been done this way for storage efficiency.

Reply:
Sorry for misleading you. What I meant is for each role having a rune. Not sure what is the first character in runes, but if take
```
RoleA = 65
RoleB = 66
...
RoleC = 85
RoleD = 86
```
Then we would encode it as string(65)+string(66)+string(85)+string(86) which is equal to "ABUV". Again we are limited but I think 100+ roles are enough.

Reply:
@arberiii We had conversation offline and switched away from bitwise design for now. @alldroll is working on a design proposal and will let us know when we can include your design in the proposal.

Reply:
I see now that it is stored as bits I thought it was as characters, my bad!